packages/babylon/src/plugins/flow.js

```diff
@@ -2219,8 +2219,56 @@ export default (superClass: Class<Parser>): Class<Parser> =>
         node.callee = base;
         node.arguments = this.parseCallExpressionArguments(tt.parenR, false);
         base = this.finishNode(node, "CallExpression");
+      } else if (
+        base.type === "Identifier" &&
+        base.name === "async" &&
+        this.isRelational("<")
+      ) {
+        const state = this.state.clone();
+        let error;
+        try {
+          const node = this.parseAsyncArrowWithTypeParameters(
+            startPos,
+            startLoc,
+          );
+          if (node) return node;
+        } catch (e) {
+          error = e;
+        }
+
+        this.state = state;
+        try {
+          return super.parseSubscripts(base, startPos, startLoc, noCalls);
+        } catch (e) {
+          throw error || e;
+        }
       }
 
       return super.parseSubscripts(base, startPos, startLoc, noCalls);
     }
+
+    parseAsyncArrowWithTypeParameters(
+      startPos: number,
+      startLoc: Position,
+    ): ?N.ArrowFunctionExpression {
+      const node = this.startNodeAt(startPos, startLoc);
+
+      const oldInFunc = this.state.inFunction;
```

Comment:
this seems really hacky, looks like the real problem is in the `parseArrow` or `parseFunctionBodyAndFinish`

Reply:
Yeah, you are right.